[wasm] Remove hardcoded references to `net8.0` from WasmApp*targets

radical opened the pull request:
Look for `System.Runtime.dll` to find the runtime pack directory with the assemblies, so we don't need to figure out the tfm for the path.
- Cannot use `System.Private.CoreLib.dll` because that is in `native/` instead of `lib/net8.0`
- Cannot use `RuntimePackAsset` because in many places that item isn't available.
  - AOT on helix where we run only the wasm targets on the proxy project.
  - Runtime tests which currently dump all the assemblies, and other bits into a single directory, not following the directory structure of a runtime pack. This also means the `RuntimePackAsset` cannot easily be synthesized.

Fixes https://github.com/dotnet/runtime/issues/79109 .

radical committed:
[wasm] Remove hardcoded references to `net8.0` from WasmApp*targets

Fixes https://github.com/dotnet/runtime/issues/79109 .

ViktorHofer commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -312,7 +312,7 @@
     <PropertyGroup>
       <_HasMscorlib Condition="'%(_WasmAssembliesInternal.FileName)%(_WasmAssembliesInternal.Extension)' == 'mscorlib.dll'">true</_HasMscorlib>
-      <_MscorlibPath Condition="'$(_HasMscorlib)' != 'true'">$(MicrosoftNetCoreAppRuntimePackRidDir)lib\net8.0\mscorlib.dll</_MscorlibPath>
+      <_MscorlibPath Condition="'$(_HasMscorlib)' != 'true'">$(MicrosoftNetCoreAppRuntimePackRidDir)lib\$(TargetFramework)\mscorlib.dll</_MscorlibPath>
```
`@(RuntimePackAsset->WithMetadataValue('Filename', 'mscorlib'))`

<img width="597" alt="image" src="https://github.com/dotnet/runtime/assets/7412651/71368cdf-0e47-457f-9bbe-52ae9fe21420">

ViktorHofer commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -628,7 +628,7 @@
 
       <_WasmAOTSearchPaths Include="$(MicrosoftNetCoreAppRuntimePackRidNativeDir)" />
-      <_WasmAOTSearchPaths Include="$(MicrosoftNetCoreAppRuntimePackRidDir)\lib\net8.0" />
+      <_WasmAOTSearchPaths Include="$(MicrosoftNetCoreAppRuntimePackRidDir)\lib\$(TargetFramework)" />
```
You could use the parent directory from a known assembly, i.e. System.Private.CoreLib or mscorlib. I would probably pick the former as at some point in the distant future, the mscorlib shim assembly could go away.

radical committed:
Address feedback from @viktorhofer

ViktorHofer commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -628,7 +628,10 @@
       <_WasmAOTSearchPaths Include="$(MicrosoftNetCoreAppRuntimePackRidNativeDir)" />
-      <_WasmAOTSearchPaths Include="$(MicrosoftNetCoreAppRuntimePackRidDir)\lib\net8.0" />
+
+      <!-- find the path with the assemblies, so we don't have to hardcode the tfm -->
+      <_SystemIOPath Include="$(MicrosoftNetCoreAppRuntimePackRidNativeDir)\lib\*\System.IO.dll" />
```
This doesn't look right. You want to save the path in a property and use the RuntimePackAsset item instead of the glob:

```
<PropertyGroup>
  <_SystemPrivateCoreLibPath>@(RuntimePackAsset->WithMetadataValue('Filename', 'System.Private.CoreLib'))</_SystemPrivateCoreLibPath>
</PropertyGroup>

<ItemGroup>
  <_WasmAOTSearchPaths Include="$([System.IO.Path]::GetDirectoryName('$(_SystemPrivateCoreLibPath)'))" />
</ItemGroup>
```

radical replied:
> This doesn't look right. You want to save the path in a property, not an item.

I was using it for the glob.

> Also, use the RuntimePackAsset item instead of the glob:

Much better idea, thanks!

radical committed:
address review feedback

ViktorHofer commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -600,6 +600,10 @@
+
+      <!-- find the path with the assemblies, so we don't have to hardcode the tfm.
+           Cannot use System.Private.Corelib since that is in a different directory -->
+      <_SystemIOPath>@(RuntimePackAsset->WithMetadataValue('Filename', 'System.IO'))</_SystemIOPath>
```
Why System.IO? System.IO is a pure shim assembly which might get trimmed out in certain scenarios.

radical replied:
Trimmed from the runtime pack?

They are in different paths:
`./Microsoft.NETCore.App.Runtime.Mono.browser-wasm/8.0.0-dev/runtimes/browser-wasm/native/System.Private.CoreLib.dll`
`./Microsoft.NETCore.App.Runtime.Mono.browser-wasm/8.0.0-dev/runtimes/browser-wasm/lib/net8.0/System.IO.dll`

.. and we need the latter one.

radical replied:
Looks like this fails for AOT/helix where we create a proxy project, and run only the wasm part of the build. And that doesn't have `RuntimePackAsset`, because none of the targets related to that were run.

I will try to populate RuntimePackAsset in the proxy project, but that will be bit of a hack.

radical committed:
Use System.Runtime.dll instead of System.IO.dll

ViktorHofer reviewed (dismissed):
If that works, looks good. Please double check that `_MscorlibPath`, `_SystemRuntimePath` and `_WasmAOTSearchPaths` include the correct values.

Thank you

radical commented:
AOT failure might be relevant. Investigating.

radical committed:
Fix local run with BuildAOTTestsOnHelix=true, to get the correct tasks config

radical committed:
Add $(MicrosoftNetCoreAppRuntimePackRidLibTfmDir)

radical committed:
WasmApp.LocalBuild.props: populate a synthetic RuntimePackAsset item, as native build depends on that

radical commented:
/azp run runtime-wasm

radical committed:
WasmApp.targets: Fix the logic to find runtime pack path with the

.. assemblies. `RuntimePackAsset` can't be reliably used because it is
not always available in the various test setups used. Instead, find
`System.Runtime.dll` in the runtime pack path, which is already known.

radical commented:
/azp run runtime-wasm

radical committed:
WasmApp.targets - error check before using the item

radical committed:
Fix runtime tests, by overriding the path to the assemblies

radical committed:
Fix runtime tests, by overriding the path to the assemblies

radical commented:
/azp run runtime-wasm

radical commented:
Failures are known, and unrelated.

pavelsavara commented:
Consider similar fix for WASI

ViktorHofer commented on src/tests/Common/wasm-test-runner/WasmTestRunner.proj:
```diff
@@ -9,6 +9,7 @@
     <EnableDefaultCompileItems>false</EnableDefaultCompileItems>
     <MicrosoftNetCoreAppRuntimePackDir>$(CORE_ROOT)\runtimepack-non-existent</MicrosoftNetCoreAppRuntimePackDir>
     <MicrosoftNetCoreAppRuntimePackRidDir>$(CORE_ROOT)\runtimepack</MicrosoftNetCoreAppRuntimePackRidDir>
+    <MicrosoftNetCoreAppRuntimePackRidLibTfmDir>$(MicrosoftNetCoreAppRuntimePackRidDir)\native</MicrosoftNetCoreAppRuntimePackRidLibTfmDir>
```
Shouldn't this be \lib\$(TFM)?

ViktorHofer replied:
Can you please file an issue for that?

radical replied:
done - https://github.com/dotnet/runtime/issues/89954

radical committed:
Apply the same changes to wasi targets also

radical committed:
fix

radical committed:
fix

radical commented:
/azp run runtime-wasm

ViktorHofer commented:
@radical you added unrelated commits, didn't you?

radical commented:
> @radical you added unrelated commits, didn't you?

The merge seems messed up a bit, so `git log` shows extra commits. *But* the only actual changes here are the relevant ones. I was avoiding cleaning this up to not mess up reviewing on github (it won't show "changes since last review" if you force-push), but I'll do that now.

radical commented:
Force pushed with the rebased commits.

ViktorHofer reviewed (approved):
Thanks
